### ISSUE TITLE
llvm package fails to install under FreeBSD

### ISSUE THREAD

**Issue by magv** [ocaml/opam-repository#8265]
Hi, folks. Here's what I get after `opam install llvm`:
```
The following actions will be performed:
  ∗  install llvm 3.9

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[llvm] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of llvm failed at "bash -ex ./build.sh 3.9
        ~/.opam/4.03.0+flambda/lib".
Processing  1/1: [llvm: sh]
#=== ERROR while installing llvm.3.9 ==========================================#
# opam-version 1.2.2
# os           freebsd
# command      bash -ex ./build.sh 3.9 ~/.opam/4.03.0+flambda/lib
# path         ~/.opam/4.03.0+flambda/build/llvm.3.9
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.env
# stdout-file  ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.out
# stderr-file  ~/.opam/4.03.0+flambda/build/llvm.3.9/llvm-13291-6d07ae.err
### stdout ###
# 77 33 p--------- 0 magv magv 0 0 "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" "/usr/local/Cellar/llvm/3.9*/bin/llvm-config" 4096 0 0 (stdin)
# Error: LLVM is not installed.
### stderr ###
# [...]
# ++ llvm-config-mp-3.9 --version
# ./build.sh: line 14: llvm-config-mp-3.9: command not found
# + continue
# + for llvm_config in 'llvm-config-$version' 'llvm-config-mp-$version' '$brew_llvm_config' llvm-config
# + case `$llvm_config --version` in
# ++ llvm-config --version
# ./build.sh: line 14: llvm-config: command not found
# + continue
# + echo 'Error: LLVM is not installed.'
# + exit 1
```

The immediate problem being that [this loop from build.sh](//github.com/ocaml/opam-repository/blob/master/packages/llvm/llvm.3.9/files/build.sh#L13):

```shell
for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
```

... does not include the actual name of `llvm-config` binary on FreeBSD, which is `llvm-config39`.

* * *

Unfortunately my attempts to fix this line on my machine and re-run the build are unsuccessful: editing `~/.opam/repo/default/packages/llvm/llvm.3.9/files/build.sh` and `~/.opam/4.03.0+flambda/build/llvm.3.9/build.sh` doesn't make a difference because `opam install` seems to ignore the first and override the second.

If anyone could advise how to actually edit a package locally to attempt to fix it, I'd appreciate that too. Creating a local repo with `opam repo add ...`, copying `~/.opam/repo/default/packages/llvm/llvm.3.9` there under a new name and trying to install that results in this error:
```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] Patch file "~/.opam/4.03.0+flambda/build/llvm-x.3.9/D24354.diff" not found.
[ERROR] Patch file "~/.opam/4.03.0+flambda/build/llvm-x.3.9/cmake.patch" not found.
```

**Comment by jpdeplaix** [comment]
One way to edit a package is to clone the opam-repository locally (```git clone https://github.com/ocaml/opam-repository```) and add it to your opam setup (```opam repository add tmp ./opam-repository```).
Now, every time you update a file locally, your just have to do ```opam update tmp``` and your changes will be taken into account.

That said, do you want me to propose a patch based on your observations or you'll just test stuffs by yourself ?

**Comment by magv** [comment]
> One way to edit a package is to clone the opam-repository locally (git clone https://github.com/ocaml/opam-repository) and add it to your opam setup (opam repository add tmp ./opam-repository).
> Now, every time you update a file locally, your just have to do opam update tmp and your changes will be taken into account.

Thanks, this does seem to work.

> That said, do you want me to propose a patch based on your observations or you'll just test stuffs by yourself ?

I don't suppose you have a FreeBSD machine handy, so I'll work on a fix and get back to you.

**Comment by magv** [comment]
OK, so here's a simple fix for llvm 3.9:
```diff
diff --git a/packages/llvm/llvm.3.9/files/build.sh b/packages/llvm/llvm.3.9/files/build.sh
index ee6dee434..d546ed4a8 100644
--- a/packages/llvm/llvm.3.9/files/build.sh
+++ b/packages/llvm/llvm.3.9/files/build.sh
@@ -10,7 +10,8 @@ fi
 
 shopt -s nullglob
 
-for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_config llvm-config; do
+version_sans_dot=$(echo $version | tr -d .)
+for llvm_config in llvm-config-$version llvm-config$version_sans_dot llvm-config-mp-$version $brew_llvm_config llvm-config; do
     case `$llvm_config --version` in
         $version|$version.*)
             case `$llvm_config --shared-mode` in
@@ -31,5 +32,5 @@ for llvm_config in llvm-config-$version llvm-config-mp-$version $brew_llvm_confi
     esac
 done
 
-echo "Error: LLVM is not installed."
+echo "Error: LLVM $version is not installed."
 exit 1 
```

This change allows llvm.3.9 package to at least detect the correct `llvm-config` binary and initiate the build.

Note that this still doesn't allow llvm.3.9 to be installed, because the latest cmake (3.7.1) will actually crash while generating makefiles. The good news is that this crash is fixed in the latest dev version of cmake, so if the change above will be committed, llvm.3.9 will become installable under FreeBSD with the next cmake release.

Here's a diff with the same change applied to all llvm versions: [opam-llvm-fbsd-fix.diff.txt](https://github.com/ocaml/opam-repository/files/713562/opam-llvm-fbsd-fix.diff.txt).
(I only tested 3.9 though).

**Comment by magv** [comment]
There's an additional problem: even if the patch above may fix the building of llvm package, using it is still impossible and results in this sort of error messages:
```
+ ocamlfind ocamlopt -linkpkg -package llvm,llvm.analysis,llvm.bitwriter,... file.cmx ... -o file.native
/usr/local/bin/ld: cannot find -lLLVMBitWriter
/usr/local/bin/ld: cannot find -lLLVMAnalysis
/usr/local/bin/ld: cannot find -lLLVMProfileData
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
/usr/local/bin/ld: cannot find -lLLVMAnalysis
/usr/local/bin/ld: cannot find -lLLVMProfileData
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
/usr/local/bin/ld: cannot find -lLLVMCore
/usr/local/bin/ld: cannot find -lLLVMSupport
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking
Command exited with code 2.
```

These libraries are all present in `/usr/local/llvm39/lib`, but it seems that this path is not added to any of the CMA files? Is there a way for the package to add this library path automatically, or do users need to specify that manually?

* * *

Finally, even if you add that patch manually (by e.g. adding ` -lflags -cclib,-L/usr/local/llvm39/lib` option to ocamlbuild), you still can't use the llvm package, because building with it results in a wall of error messages like ``undefined reference to `std::__1::system_category()'``; adding `-lflags -cclib,-lc++` to ocamlbuild arguments fixes this. Again, is there a way for the package to do this automatically?

**Comment by jpdeplaix** [comment]
Have you tried to add ```-linkall``` to the ocamlfind command ? Are you using ocamlbuild ?

**Comment by magv** [comment]
Yes, I'm using ocamlbuild. Adding `-lflags -linkall` to ocamlbuild arguments still results in the same error:
```
$ ocamlbuild -no-links -use-ocamlfind -r \
    -lflags -linkall \
    -tag-line "<*>: package(llvm,llvm.analysis,llvm.bitwriter)"
+ ocamlfind ocamlopt -linkall -linkpkg -package llvm,...
/usr/local/bin/ld: cannot find -lLLVMBitWriter
/usr/local/bin/ld: cannot find -lLLVMAnalysis
...
```

**Comment by jpdeplaix** [comment]
Could you show what is the output of ```llvm-config39 --shared-mode``` please ?
If it is "static", could you try to apply the ```link.patch``` patch also in this case (see build.sh).
If not, I really don't understand.

In any cases, poke @whitequark.

**Comment by whitequark** [comment]
@magv What's the list of files installed by the FreeBSD LLVM package?

**Comment by magv** [comment]
@jpdeplaix, `llvm-config39 --shared-mode` is static. If I force `link.patch` application the error changes to `/usr/local/bin/ld: cannot find -lLLVM-3.9`.

@whitequark, here's the list: [llvm39.files.txt](https://github.com/ocaml/opam-repository/files/714532/llvm39.files.txt); you can also more or less see it in `pkg-plist` file [over here](https://svnweb.freebsd.org/ports/head/devel/llvm39/).

**Comment by magv** [comment]
So, any news, folks?

Can we at least commit the patch I proposed a few comments back? It fixes the installation, even if the library path problem remains.

**Comment by jpdeplaix** [comment]
Sorry for the delay, I'm not really experienced for this kind of problems but at least I opened a PR using your patch

**Comment by jpdeplaix** [comment]
Mmh it's only partially fixed. Does somebody can reopen this issue ?

**Comment by dbuenzli** [comment]
This seems to have finally have been fixed by #8700 and #8753 and I think this can be closed.
